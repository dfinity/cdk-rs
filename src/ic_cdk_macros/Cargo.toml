[package]
name = "ic_cdk_macros"
version = "0.1.0"
authors = ["Hans Larsen <hans@larsen.online>"]
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[lib]
proc-macro = true

[dependencies]
candid = { git = "https://github.com/hansl/candid", branch = "rust-cdk" }
syn = { version = "1.0", features = ["fold", "full"] }
quote = "1.0"
proc-macro2 = "1.0"
