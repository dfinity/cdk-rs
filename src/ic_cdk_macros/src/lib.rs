use proc_macro::TokenStream;
use std::sync::atomic::{AtomicU32, Ordering};

mod error;
mod export;
mod import;

static NEXT_ID: AtomicU32 = AtomicU32::new(0);
pub(crate) fn id() -> u32 {
    NEXT_ID.fetch_add(1, Ordering::SeqCst)
}

fn handle_debug_and_errors<F>(
    cb: F,
    name: &str,
    attr: TokenStream,
    item: TokenStream,
) -> TokenStream
where
    F: FnOnce(TokenStream, TokenStream) -> Result<TokenStream, error::Errors>,
{
    if std::env::var_os("IC_CDK_DEBUG").is_some() {
        eprintln!("--- IC_CDK_MACROS DEBUG ---");
        eprintln!("{}\n    attr: {}\n    item: {}", name, attr, item);
    }

    let result = cb(attr, item);

    if std::env::var_os("IC_CDK_DEBUG").is_some() {
        eprintln!("--------- RESULT  ---------");
        if let Ok(ref stream) = result {
            eprintln!("{}", stream);
        }
        eprintln!("---------------------------");
    }

    // We return an empty tokenstream on a fatal error as not to display the error
    // twice.
    let (errors, res) = result.map_or_else(
        |e| (e, TokenStream::new()),
        |v| (error::Errors::default(), v),
    );

    errors.emit();
    res
}

#[proc_macro_attribute]
pub fn query(attr: TokenStream, item: TokenStream) -> TokenStream {
    handle_debug_and_errors(export::ic_query, "ic_query", attr, item)
}

#[proc_macro_attribute]
pub fn update(attr: TokenStream, item: TokenStream) -> TokenStream {
    handle_debug_and_errors(export::ic_update, "ic_update", attr, item)
}

#[proc_macro_attribute]
pub fn init(attr: TokenStream, item: TokenStream) -> TokenStream {
    handle_debug_and_errors(export::ic_init, "ic_init", attr, item)
}

#[proc_macro_attribute]
pub fn import(attr: TokenStream, item: TokenStream) -> TokenStream {
    handle_debug_and_errors(import::ic_import, "ic_import", attr, item)
}
