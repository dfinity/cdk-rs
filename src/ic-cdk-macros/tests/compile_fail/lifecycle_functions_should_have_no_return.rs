use ic_cdk_macros::*;

#[init]
fn init() -> u32 {}

#[pre_upgrade]
fn pre_upgrade() -> u32 {}

#[post_upgrade]
fn post_upgrade() -> u32 {}

#[heartbeat]
fn heartbeat() -> u32 {}

#[inspect_message]
fn inspect_message() -> u32 {}

fn main() {}
