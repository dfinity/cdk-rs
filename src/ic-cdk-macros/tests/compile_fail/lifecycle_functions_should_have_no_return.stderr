error: #[init] function cannot have a return value.
 --> tests/compile_fail/lifecycle_functions_should_have_no_return.rs:3:1
  |
3 | #[init]
  | ^^^^^^^
  |
  = note: this error originates in the attribute macro `init` (in Nightly builds, run with -Z macro-backtrace for more info)

error: #[pre_upgrade] function cannot have a return value.
 --> tests/compile_fail/lifecycle_functions_should_have_no_return.rs:6:1
  |
6 | #[pre_upgrade]
  | ^^^^^^^^^^^^^^
  |
  = note: this error originates in the attribute macro `pre_upgrade` (in Nightly builds, run with -Z macro-backtrace for more info)

error: #[post_upgrade] function cannot have a return value.
 --> tests/compile_fail/lifecycle_functions_should_have_no_return.rs:9:1
  |
9 | #[post_upgrade]
  | ^^^^^^^^^^^^^^^
  |
  = note: this error originates in the attribute macro `post_upgrade` (in Nightly builds, run with -Z macro-backtrace for more info)

error: #[heartbeat] function cannot have a return value.
  --> tests/compile_fail/lifecycle_functions_should_have_no_return.rs:12:1
   |
12 | #[heartbeat]
   | ^^^^^^^^^^^^
   |
   = note: this error originates in the attribute macro `heartbeat` (in Nightly builds, run with -Z macro-backtrace for more info)

error: #[inspect_message] function cannot have a return value.
  --> tests/compile_fail/lifecycle_functions_should_have_no_return.rs:15:1
   |
15 | #[inspect_message]
   | ^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the attribute macro `inspect_message` (in Nightly builds, run with -Z macro-backtrace for more info)
